File: src/main/resources/generators/ts/native/tst.ts
export class DateWrapper {
    private inner: Date;

    constructor(inner: Date) {
        this.inner = inner;
    }

    get timestamp() {
        return new NumberWrapper(this.inner.getTime());
    }

    get day() {
        return new NumberWrapper(this.inner.getDate());
    }

    get month() {
        return new NumberWrapper(this.inner.getMonth());
    }

    toDate() {
        return this.inner;
    }

    equals(date: DateWrapper): boolean {
        return this.timestamp == date.timestamp;
    }

    notEquals(date: DateWrapper): boolean {
        return this.timestamp != date.timestamp;
    }

    upper(date: DateWrapper): boolean {
        return this.timestamp > date.timestamp;
    }

    lower(date: DateWrapper): boolean {
        return this.timestamp < date.timestamp;
    }

    upperOrEquals(date: DateWrapper): boolean {
        return this.timestamp >= date.timestamp;
    }

    lowerOrEquals(date: DateWrapper): boolean {
        return this.timestamp <= date.timestamp;
    }
}
export class NumberWrapper {
    private inner: Number;

    constructor(inner: Number) {
        this.inner = inner;
    }

    toNumber() {
        return this.inner;
    }

    minus(numberWrapper: NumberWrapper): NumberWrapper {
        return new NumberWrapper(this.inner - numberWrapper.inner);
    }

    plus(numberWrapper: NumberWrapper): NumberWrapper {
        return new NumberWrapper(this.inner + numberWrapper.inner);
    }

    time(numberWrapper: NumberWrapper): NumberWrapper {
        return new NumberWrapper(this.inner * numberWrapper.inner);
    }

    divide(numberWrapper: NumberWrapper): NumberWrapper {
        return new NumberWrapper(this.inner / numberWrapper.inner);
    }

    modulo(numberWrapper: NumberWrapper): NumberWrapper {
        return new NumberWrapper(this.inner % numberWrapper.inner);
    }

    equals(numberWrapper: NumberWrapper): boolean {
        return this.inner === numberWrapper.inner;
    }

    notEquals(numberWrapper: NumberWrapper): boolean {
        return this.inner !== numberWrapper.inner;
    }

    upper(numberWrapper: NumberWrapper): boolean {
        return this.inner > numberWrapper.inner;
    }

    lower(numberWrapper: NumberWrapper): boolean {
        return this.inner < numberWrapper.inner;
    }

    upperOrEquals(numberWrapper: NumberWrapper): boolean {
        return this.inner >= numberWrapper.inner;
    }

    lowerOrEquals(numberWrapper: NumberWrapper): boolean {
        return this.inner <= numberWrapper.inner;
    }
}
export class StringWrapper{
    private inner: string;

    constructor(inner: string) {
        this.inner = inner || "";
    }

    toString() {
        return this.inner;
    }

    nonEmpty(): boolean {
        return this.inner.length > 0;
    }

    trim(): StringWrapper {
        return new StringWrapper(this.inner.trim());
    }

    startsWith(prefix: StringWrapper): boolean {
        return this.inner.substr(0, prefix.inner.length) === prefix.inner;
    }

    matches(regex: StringWrapper): boolean {
        return new RegExp(regex.inner).test(this.inner);
    }
}



export function verifyNonEmpty(string: StringWrapper): string|null {
    return verify("The string is empty // quoted comment", () => {
        return (string).nonEmpty();
    })
}



export function verifyNonBlank(string: StringWrapper): string|null {
    return verify("The string is blank /* quoted comment */", () => {
        return ((string).trim()).nonEmpty();
    })
}


/*
 Could be simplified, but it is for the "example"
 */
export function verifyPhoneNumber(string: StringWrapper): string|null {
    return verify("Please provide a phone number", () => {
        return  ((string).nonEmpty())
            ? ( ((string).startsWith(new StringWrapper("+33")))
                    ? ((string).matches(new StringWrapper("^\\+33\\d{9}$")))
                    : ((string).matches(new StringWrapper("^0\\d{9}$")))
            )
            : (false)
            ;
    })
}



export function verifyYearPeriod(period: $Period): string|null {
    return verify("The period must last one year", () => {
        return ((((period).end).timestamp).minus(((period).start).timestamp)) == ((new NumberWrapper(365)).time((new NumberWrapper(24)).time(new NumberWrapper(3600))));
    })
}



export function verifyStartJanuaryFirst(period: $Period): string|null {
    return verify("The period must start on january the first", () => {
        return ((((period).start).day).equals(new NumberWrapper(1))) && ((((period).start).month).equals(new NumberWrapper(1)));
    })
}


function verify(message: string, condition: () => Boolean): string|null {
    if (condition()) {
        return null;
    } else {
        return message;
    }
}



interface $Period {
    start: DateWrapper;
    end: DateWrapper;
}


export interface Period extends $Period {}

export function Period(start: DateWrapper, end: DateWrapper): string|Readonly<Period> {
    const __result = {start, end};
    const period = __result;
    let __errorOpt = null;
    const __verifications = [

        verify("end should be after start", () => {
            return (((period).end).upper((period).start)) || (((period).end).equals((period).start));
        })

    ];
    for (let __i = 0 ; __i < __verifications.length ; __i++) {
        if (__verifications[__i] !== null) {
            __errorOpt = __verifications[__i];
        }
    }

    if (__errorOpt) {
        return __errorOpt;
    } else {
        return Object.freeze(__result);
    }
}




export interface CivilYear extends $Period {}

export function CivilYear(start: DateWrapper, end: DateWrapper): string|Readonly<CivilYear> {
    const __result = {start, end};
    const period = __result;
    let __errorOpt = null;
    const __verifications = [

        verify("end should be after start", () => {
            return (((period).end).upper((period).start)) || (((period).end).equals((period).start));
        })
        , verifyYearPeriod(__result), verifyStartJanuaryFirst(__result)
    ];
    for (let __i = 0 ; __i < __verifications.length ; __i++) {
        if (__verifications[__i] !== null) {
            __errorOpt = __verifications[__i];
        }
    }

    if (__errorOpt) {
        return __errorOpt;
    } else {
        return Object.freeze(__result);
    }
}

const y = CivilYear(null, null);
if ((y as CivilYear).start) {

} else {

}

File: src/main/resources/generators/scala/native/tst.scala
package object booleanWrapperUtils {
  implicit def booleanToBooleanWrapper(boolean: Boolean): BooleanWrapper = new BooleanWrapper(boolean)
  implicit def booleanWrapperToBoolean(booleanWrapper: BooleanWrapper): Boolean = booleanWrapper.inner
}

import booleanWrapperUtils._

class BooleanWrapper(val inner: Boolean) {

}
import java.util.Date

// TODO: Use newer Date API
class DateWrapper(private val inner: Date) {
  def timestamp: NumberWrapper = new NumberWrapper(inner.getTime())

  def day: NumberWrapper = new NumberWrapper(inner.getDate)

  def month: NumberWrapper = new NumberWrapper(inner.getMonth)

  def ==(dateWrapper: DateWrapper): BooleanWrapper = timestamp == dateWrapper.timestamp

  def !=(dateWrapper: DateWrapper): BooleanWrapper = timestamp != dateWrapper.timestamp

  def >(dateWrapper: DateWrapper): BooleanWrapper = timestamp > dateWrapper.timestamp

  def <(dateWrapper: DateWrapper): BooleanWrapper = timestamp < dateWrapper.timestamp

  def >=(dateWrapper: DateWrapper): BooleanWrapper = timestamp >= dateWrapper.timestamp

  def <=(dateWrapper: DateWrapper): BooleanWrapper = timestamp <= dateWrapper.timestamp
}
class NumberWrapper(private val inner: BigDecimal) {
  def -(numberWrapper: NumberWrapper): NumberWrapper = new NumberWrapper(inner - numberWrapper.inner)

  def +(numberWrapper: NumberWrapper): NumberWrapper = new NumberWrapper(inner + numberWrapper.inner)

  def *(numberWrapper: NumberWrapper): NumberWrapper = new NumberWrapper(inner * numberWrapper.inner)

  def /(numberWrapper: NumberWrapper): NumberWrapper = new NumberWrapper(inner / numberWrapper.inner)

  def %(numberWrapper: NumberWrapper): NumberWrapper = new NumberWrapper(inner % numberWrapper.inner)

  def ==(numberWrapper: NumberWrapper): BooleanWrapper = inner == numberWrapper.inner

  def !=(numberWrapper: NumberWrapper): BooleanWrapper = inner != numberWrapper.inner

  def >(numberWrapper: NumberWrapper): BooleanWrapper = inner > numberWrapper.inner

  def <(numberWrapper: NumberWrapper): BooleanWrapper = inner < numberWrapper.inner

  def >=(numberWrapper: NumberWrapper): BooleanWrapper = inner >= numberWrapper.inner

  def <=(numberWrapper: NumberWrapper): BooleanWrapper = inner <= numberWrapper.inner
}
class StringWrapper(private val inner: String) {
  def nonEmpty(): BooleanWrapper = inner.nonEmpty

  def trim(): StringWrapper = new StringWrapper(inner.trim)

  def startsWith(prefix: StringWrapper): BooleanWrapper = inner.startsWith(prefix.inner)

  def matches(regex: StringWrapper): BooleanWrapper = inner.matches(regex.inner)
}

package object verifications {


  def verifyNonEmpty(string: StringWrapper): Option[String] = {
    verify("The string is empty // quoted comment") {
      (string).nonEmpty()
    }
  }




  def verifyNonBlank(string: StringWrapper): Option[String] = {
    verify("The string is blank /* quoted comment */") {
      ((string).trim()).nonEmpty()
    }
  }



  /*
    Could be simplified, but it is for the "example"
  */
  def verifyPhoneNumber(string: StringWrapper): Option[String] = {
    verify("Please provide a phone number") {

      if ((string).nonEmpty()) {

        if ((string).startsWith(new StringWrapper("+33"))) {
          (string).matches(new StringWrapper("^\\+33\\d{9}$"))
        } else {
          (string).matches(new StringWrapper("^0\\d{9}$"))
        }

      } else {
        new BooleanWrapper(false)
      }

    }
  }




  def verifyYearPeriod(period: $Period): Option[String] = {
    verify("The period must last one year") {
      ((((period).end).timestamp) - (((period).start).timestamp)) == ((new NumberWrapper(365)) * ((new NumberWrapper(24)) * (new NumberWrapper(3600))))
    }
  }




  def verifyStartJanuaryFirst(period: $Period): Option[String] = {
    verify("The period must start on january the first") {
      ((((period).start).day) == (new NumberWrapper(1))) && ((((period).start).month) == (new NumberWrapper(1)))
    }
  }


  private def verify(message: String)(condition: => Boolean) = {
    if (condition) {
      None
    } else {
      Some(message)
    }
  }
}
import verifications._



trait $Period {
  def start: DateWrapper
  def end: DateWrapper
}


class Period private(val start: DateWrapper, val end: DateWrapper) extends $Period

object Period {
  def apply(start: DateWrapper, end: DateWrapper): Either[String, Period] = {
    val __result = new Period(start, end)
    val period = __result
    val __errorOpt = Seq(

      verify("end should be after start") {
        (((period).end) > ((period).start)) || (((period).end) == ((period).start))
      }

    ).find(_.isDefined).flatten

    __errorOpt match {
      case Some(error) => Left(error)
      case None =>

        Right(__result)
    }
  }

  private def verify(message: String)(condition: => Boolean) = {
    if (condition) {
      None
    } else {
      Some(message)
    }
  }
}





class CivilYear private(val start: DateWrapper, val end: DateWrapper) extends $Period

object CivilYear {
  def apply(start: DateWrapper, end: DateWrapper): Either[String, CivilYear] = {
    val __result = new CivilYear(start, end)
    val period = __result
    val __errorOpt = Seq(

      verify("end should be after start") {
        (((period).end) > ((period).start)) || (((period).end) == ((period).start))
      }
      , verifyYearPeriod(__result), verifyStartJanuaryFirst(__result)
    ).find(_.isDefined).flatten

    __errorOpt match {
      case Some(error) => Left(error)
      case None =>

        Right(new CivilYear(__result.start, __result.end))
    }
  }

  private def verify(message: String)(condition: => Boolean) = {
    if (condition) {
      None
    } else {
      Some(message)
    }
  }
}